### PR TITLE
[Linter] Update regex to find resource, child resource and grandchild resource names from path

### PR DESCRIPTION
**Pull request by vishrutshah**
Fixes https://github.com/Azure/autorest/issues/1988

Example:
```
{
    "type": "Error",
    "code": "OperationIdNounInVerb",
    "message": "Per the Noun_Verb convention for Operation Ids, the noun 'Recommendations' should not appear after the underscore.",
    "jsonref": "file:///D:/Repos/azure-rest-api-specs/arm-advisor/2016-07-12-preview/swagger/advisor.json#/paths/~1subscriptions~1{subscriptionId}~1resourceGroups~1{resourceGroupName}~1providers~1Microsoft.Sql~1servers~1{serverName}~1databases~1{databaseName}~1pause/get/operationId",
    "json-path": "file:///D:/Repos/azure-rest-api-specs/arm-advisor/2016-07-12-preview/swagger/advisor.json#/paths/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/databases/{databaseName}/pause/get/operationId",
    "id": "M1001",
    "validationCategory": "SDKViolation",
    "providerNamespace": "Microsoft.Sql",
    "resourceType": "databases"
  },
  {
    "type": "Error",
    "code": "BodyTopLevelProperties",
    "message": "Top level properties should be one of name, type, id, location, properties, tags, plan, sku, etag, managedBy, identity. Extra properties found: \"SuppressionContract/suppressionId, SuppressionContract/ttl\".",
    "jsonref": "file:///D:/Repos/azure-rest-api-specs/arm-advisor/2016-07-12-preview/swagger/advisor.json#/paths/~1{resourceUri}~1providers~1Microsoft.Advisor~1recommendations~1{recommendationId}~1suppressions~1{name}",
    "json-path": "file:///D:/Repos/azure-rest-api-specs/arm-advisor/2016-07-12-preview/swagger/advisor.json#/paths/{resourceUri}/providers/Microsoft.Advisor/recommendations/{recommendationId}/suppressions/{name}",
    "id": "M3006",
    "validationCategory": "RPCViolation",
    "providerNamespace": "Microsoft.Advisor",
    "resourceType": "suppressions"
  }
```

### CODE REVIEW

File: src/modeler/AutoRest.Swagger/Validation/Core/JsonValidationLogListener.cs [hunk 1]
```diff
@@ -29,7 +29,23 @@ public void Log(LogMessage message)
                     .FirstOrDefault();
                 var pathComponents = resPathPattern.Match(path ?? "");
                 var pathComponentProviderNamespace = pathComponents.Groups["providerNamespace"];
-                var pathComponentResourceType = pathComponents.Groups["resourceType"];
+                var pathComponentResource = pathComponents.Groups["resource"];
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/Core/JsonValidationLogListener.cs]: Or, use the same regex and check the number of groups returned by regex and accordingly assign, parent, child or grandchild

**Comment by vishrutshah** [comment]
@dsgouda The idea here is to find the correct type that is violating this rule. That'd help @ravbhatnagar to the reporting at the resource levels as resources/childresource can be owned by different teams for one service. 

Nope I think we cannot be sure on number of levels for nesting. I am also not sure if we can have regex to match any level. 

@dsgouda Do you have any suggestion that can help us finding this at any level?
@ravbhatnagar Do you know upto what level we would like to extract this information?

**Comment by dsgouda** [comment]
Theoretically you could apply the same regex to pick as many matches as possible and start reading them last match first which should give you the deepest node.
However, is that what we want? I wasn't a part of this discussion and I suggest we refine this requirement as best as we can. Please talk to @amarzavery too

**Comment by vishrutshah** [comment]
Here is an example on checked-in sql swagger:

```
 {
    "type": "Error",
    "code": "BodyTopLevelProperties",
    "message": "Top level properties should be one of name, type, id, location, properties, tags, plan, sku, etag, managedBy, identity. Extra properties found: \"Database/kind\".",
    "jsonref": "https://raw.githubusercontent.com/Azure/azure-rest-api-specs/master/arm-sql/2014-04-01/swagger/sql.core.json#/paths/~1subscriptions~1{subscriptionId}~1resourceGroups~1{resourceGroupName}~1providers~1Microsoft.Sql~1servers~1{serverName}~1databases~1{databaseName}",
    "json-path": "https://raw.githubusercontent.com/Azure/azure-rest-api-specs/master/arm-sql/2014-04-01/swagger/sql.core.json#/paths/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/databases/{databaseName}",
    "id": "M3006",
    "validationCategory": "RPCViolation",
    "providerNamespace": "Microsoft.Sql",
    "resourceType": "databases"
  },
  {
    "type": "Error",
    "code": "BodyTopLevelProperties",
    "message": "Top level properties should be one of name, type, id, location, properties, tags, plan, sku, etag, managedBy, identity. Extra properties found: \"Server/kind\".",
    "jsonref": "https://raw.githubusercontent.com/Azure/azure-rest-api-specs/master/arm-sql/2014-04-01/swagger/sql.core.json#/paths/~1subscriptions~1{subscriptionId}~1resourceGroups~1{resourceGroupName}~1providers~1Microsoft.Sql~1servers~1{serverName}",
    "json-path": "https://raw.githubusercontent.com/Azure/azure-rest-api-specs/master/arm-sql/2014-04-01/swagger/sql.core.json#/paths/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}",
    "id": "M3006",
    "validationCategory": "RPCViolation",
    "providerNamespace": "Microsoft.Sql",
    "resourceType": "servers"
  },
```

**Comment by vishrutshah** [comment]
@veronicagg I updated the regex to compare & group any level and pick the last one.

Here is the sample output for compute:

```
  {
    "type": "Info",
    "code": "AvoidAnonymousTypes",
    "message": "For better generated code quality, define schemas instead of using anonymous types.",
    "jsonref": "file:///D:/Repos/azure-rest-api-specs/arm-compute/2016-04-30-preview/swagger/compute.json#/paths/~1subscriptions~1{subscriptionId}~1providers~1Microsoft.Compute~1locations~1{location}~1publishers~1{publisherName}~1artifacttypes~1vmimage~1offers~1{offer}~1skus~1{skus}~1versions/get/responses/200/schema",
    "json-path": "file:///D:/Repos/azure-rest-api-specs/arm-compute/2016-04-30-preview/swagger/compute.json#/paths/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/publishers/{publisherName}/artifacttypes/vmimage/offers/{offer}/skus/{skus}/versions/get/responses/200/schema",
    "id": "!!! implement me and make me abstract !!!",
    "validationCategory": "None",
    "providerNamespace": "Microsoft.Compute",
    "resourceType": "versions"
  },
  {
    "type": "Info",
    "code": "AvoidAnonymousTypes",
    "message": "For better generated code quality, define schemas instead of using anonymous types.",
    "jsonref": "file:///D:/Repos/azure-rest-api-specs/arm-compute/2016-04-30-preview/swagger/compute.json#/paths/~1subscriptions~1{subscriptionId}~1providers~1Microsoft.Compute~1locations~1{location}~1publishers~1{publisherName}~1artifacttypes~1vmimage~1offers/get/responses/200/schema",
    "json-path": "file:///D:/Repos/azure-rest-api-specs/arm-compute/2016-04-30-preview/swagger/compute.json#/paths/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/publishers/{publisherName}/artifacttypes/vmimage/offers/get/responses/200/schema",
    "id": "!!! implement me and make me abstract !!!",
    "validationCategory": "None",
    "providerNamespace": "Microsoft.Compute",
    "resourceType": "offers"
  },
  {
    "type": "Info",
    "code": "AvoidAnonymousTypes",
    "message": "For better generated code quality, define schemas instead of using anonymous types.",
    "jsonref": "file:///D:/Repos/azure-rest-api-specs/arm-compute/2016-04-30-preview/swagger/compute.json#/paths/~1subscriptions~1{subscriptionId}~1providers~1Microsoft.Compute~1locations~1{location}~1publishers/get/responses/200/schema",
    "json-path": "file:///D:/Repos/azure-rest-api-specs/arm-compute/2016-04-30-preview/swagger/compute.json#/paths/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/publishers/get/responses/200/schema",
    "id": "!!! implement me and make me abstract !!!",
    "validationCategory": "None",
    "providerNamespace": "Microsoft.Compute",
    "resourceType": "publishers"
  },
  {
    "type": "Info",
    "code": "AvoidAnonymousTypes",
    "message": "For better generated code quality, define schemas instead of using anonymous types.",
    "jsonref": "file:///D:/Repos/azure-rest-api-specs/arm-compute/2016-04-30-preview/swagger/compute.json#/paths/~1subscriptions~1{subscriptionId}~1providers~1Microsoft.Compute~1locations~1{location}~1publishers~1{publisherName}~1artifacttypes~1vmimage~1offers~1{offer}~1skus/get/responses/200/schema",
    "json-path": "file:///D:/Repos/azure-rest-api-specs/arm-compute/2016-04-30-preview/swagger/compute.json#/paths/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/publishers/{publisherName}/artifacttypes/vmimage/offers/{offer}/skus/get/responses/200/schema",
    "id": "!!! implement me and make me abstract !!!",
    "validationCategory": "None",
    "providerNamespace": "Microsoft.Compute",
    "resourceType": "skus"
  },
```

File: src/modeler/AutoRest.Swagger/Validation/Core/JsonValidationLogListener.cs [hunk 2]
```diff
@@ -14,7 +14,7 @@ public class JsonValidationLogListener : ILogListener
     {
         private readonly List<Dictionary<string, string>> rawMessageCollection = new List<Dictionary<string, string>>();
 
-        private readonly Regex resPathPattern = new Regex(@"/providers/(?<providerNamespace>[^{/]+)/((?<resourceType>[^{/]+)/)?");
+        private readonly Regex resPathPattern = new Regex(@"/providers/(?<providerNamespace>[^{/]+)((/(?<resource>[^{/]+)/)((?<resourceName>[^/]+)))+(/(?<resource>[^{/]+))");
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/Core/JsonValidationLogListener.cs]: nit: mark this as static so we don't have to generate this object for every validation

**Comment by vishrutshah** [comment]
Thanks for the review!!

**Pull request merged**